src/dev/HallSensor.cpp

```diff
@@ -0,0 +1,112 @@
+constexpr uint32_t THRESHOLD = 5000;// Threshold for wheel speed
+
+namespace WSS::DEV {
+
+HallSensor::HallSensor(IO::GPIO& gpio, uint32_t wheelRadius, uint32_t numberOfMagnets) : gpio((IO::GPIO&) gpio), wheelRadius(wheelRadius), numberOfMagnets(numberOfMagnets) {
+    this->prevTime = 0;
+    this->lastInterval = 0;
+    this->state = WheelSpeedState::STOP;
+    this->magnetInLastRead = false;
+}
+
+void HallSensor::update() {
+    uint32_t now = HAL_GetTick();
+    uint32_t elapsedTime = now - prevTime;
+
+    // The magnet is detected if the previous gpio pin reading is the same as the current gpio
+    // pin reading
+    const bool magnetDetected = gpio.readPin() == MAGNET_DETECTED_STATE;
+    bool magnetDetectedNow = false;
+    if (!magnetWasDetected && magnetDetected != magnetInLastRead && magnetDetected) {
+        magnetWasDetected = true;
+        magnetDetectedNow = true;
+    } else if (magnetWasDetected && magnetDetected == magnetInLastRead) {
+        magnetWasDetected = false;
+    }
+    magnetInLastRead = magnetDetected;
+
+    switch (state) {
+    case WheelSpeedState::STOP:
+        if (magnetDetectedNow) {
+            prevTime = now;
+            state = WheelSpeedState::INITIALIZING;
+        }
+        break;
+
+    case WheelSpeedState::INITIALIZING:
+        if (magnetDetectedNow) {
+            // If the magnet is detected for the second time, record the time and the first
+            // interval and enter the maintain state
+            lastInterval = elapsedTime;
+            prevTime = now;
+            state = WheelSpeedState::MAINTAIN;
+        }
+
+        // If the magnet hasn't been detected for a significant amount of time, go back to the stop
+        // state
+        if (elapsedTime > THRESHOLD) {
+            state = WheelSpeedState::STOP;
+            prevTime = 0;
+        }
+        break;
+
+    case WheelSpeedState::MAINTAIN:
+        if (magnetDetectedNow) {
+            // If the magnet is newly detected, record the time and interval
+            lastInterval = elapsedTime;
+            prevTime = now;
+        }
+
+        // If the magnet hasn't been detected for a significant amount of time, go back to the stop
+        // state
+        if (elapsedTime > THRESHOLD) {
+            state = WheelSpeedState::STOP;
+            prevTime = 0;
+            lastInterval = 0;
+            // If the elapsed time since the magnet was last detected exceeds the last interval, record
+            // the last interval as if the magnet was just detected because the bike is slowing
+            // down, but the exact speed can't be calculated
+        } else if (elapsedTime > lastInterval) {
+            lastInterval = elapsedTime;
+        }
+        break;
+    }
+}
+
+uint32_t HallSensor::getRawInterval() {
+    return lastInterval;
+}
+
+uint32_t HallSensor::getSpeed() {
+    if (state == WheelSpeedState::STOP) {
+        // The wheel isn't moving, so its speed is zero.
+        return 0;
+    }
```

Comment:
We should also return 0 if the state is `INITIALIZING` because we don't have any way to give a realistic estimate of the speed yet